rfd/0206-sigstore-workload-attestation.md

```diff
@@ -0,0 +1,565 @@
+
+These components and their roles are discussed in detail below.
+
+#### Key Management
+
+Securely distributing and managing the lifecycle of signing keys is one of the
+hardest problems to solve when designing a code signing solution. They're
+essentially highly sensitive and long-lived secrets, which are problematic for
+reasons we're all familiar with!
+
+Sigstore takes a novel approach to this problem by binding signatures to an OIDC
+**identity** rather than a long-lived key. It does so by running a certificate
+authority called Fulcio, which exchanges an OIDC identity token for a
+short-lived signing certificate which is used once and then immediately
+discarded.
+
+When using the Cosign CLI to sign an image locally, the user simply runs:
+
+```shell
+$ cosign sign <image>@sha256:<hash>
+```
+
+Which opens a browser window for them to log in with their Google, Microsoft, or
+GitHub account, signs the artifact, and uploads the signature to the container
+registry.
+
+When running `cosign` non-interactively in CI platforms, etc. you can pass the
+identity token (i.e. your GitHub Actions OIDC token) as a command line flag.
+
+On the receiving end, users can verify image signatures using the Cosign CLI
+like so:
+
+```shell
+$ cosign verify <image>@sha256:<hash> \
+  --certificate-identity daniel.upton@goteleport.com \
+  --certificate-oidc-issuer https://accounts.google.com
+```
+
+There is also a Kubernetes [admission controller](https://docs.sigstore.dev/policy-controller/overview/)
+that can be used to only allow images signed by trusted authorities to run in
+your cluster. While this is conceptually similar to the feature described by
+this RFD, we think of the two as complementary rather than competing, as using
+both would allow for a "defense in depth" approach and because organizations may
+have multiple Kubernetes clusters operated by different groups.
+
+#### Transparency Log
+
+Cosign also records the signature in Rekor, the signature transparency log. This
+solves a number of problems, including:
+
+1. Because the certificates Fulcio issues are so short-lived (default 10 minutes)
+   they will almost certainly have expired by the time the artifact signature is
+   verified. As such, you need another way to prove the certificate was valid at
+   the time the signature was generated. Rekor solves this by, in essence,
+   timestamping the signature in its transparency log.
+
+2. In the same way that a certificate transparency log allows you to monitor for
+   abuse of your organization's domains, and discover if the CA itself has been
+   compromised; the signature transparency log enables you to find out if an
+   unauthorized party is signing artifacts on your behalf, or if the Sigstore
+   infrastructure has been compromised.
+
+You can also opt to counter-sign the signature using your own RFC 3161 timestamp
+authority by passing the `--timestamp-server-url` flag.
+
+#### Public Good Infrastructure
```

Comment:
Are there any rate limits around Fulcio/Rekor we should be aware of? Do we need to implement caching to avoid these rate limits?

Reply:
Great question! While there aren't any publicly communicated rate limits, there clearly are some (I assume IP-based) limits to prevent abuse: https://github.com/sigstore/cosign/issues/3153.

Maybe could implement an LRU cache where the key is `SHA-256(SigstorePolicy) + "." + SHA-256(bundle)`, for clients that refresh their SVIDs frequently?

Reply:
I'll leave this one to your gut feel - imho - if there's no published numbers then it's probably fairly high and unlikely to be an issue for most initial adopters. If you think the effort isn't worth the risk at this point, then perhaps include mention of LRU in the RFD as a future improvement.